### PR TITLE
Predict uncertainty in outputs

### PR DESCRIPTION
**Pull request by peastman**
Implements the feature described in #1119.

So far I've created the basic architecture and implemented uncertainty for MultitaskRegressor.  Let me know if the design looks ok.

I'm not sure how to implement this for classifiers.  In the paper they add Gaussian noise to the logits, then use Monte Carlo integration to determine the uncertainty in the probabilities.  And they do that during training as well as prediction.  To compute the loss at each training step, they sample many different noise vectors, pass each one through the softmax, compute the loss for every one, then average over them to get the final loss.  I'm not sure how to fit that into our architecture.  For the moment, maybe we should only support uncertainty for regression models.

Also note that unlike `predict()`, `predict_uncertainty()` does not let you specify transformers.  That's because we don't know how to propagate the uncertainty through the transforms.

### CODE REVIEW

**Comment by rbharath** [comment]
@peastman I really like the basic design. Provides a simple clean API for estimating uncertainty for regression models. I agree that the classifier uncertainty is a little trickier, and I think proceeding with good regression uncertainty support to start sounds like a solid start.

Good to merge in once you feel this is ready.

**Comment by peastman** [comment]
Ok.  I'll try implementing a few more models just to make sure I haven't made any assumptions that don't generalize.  And I'll write a tutorial on computing uncertainty.

**Comment by peastman** [comment]
`GraphConvModel` *already* has a method for predicting uncertainty???  It's called `bayesian_predict()`.  That seems to be the only model with a method of that name.  According to the docstring, it's based on the earlier paper by the same author, so it includes only epistemic uncertainty (calculated with dropout), not aleatoric uncertainty.

Shall I just get rid of that method?  @patrickhop it looks like this was your code.

**Comment by peastman** [comment]
`GraphConvModel` only makes very limited use of dropout.  It doesn't include any dropout after the convolutional layers.  It only uses it in two places:

1. After all the convolution layers, there is a single dense layer.  That is followed by a single dropout layer.
2. That is sent through a GraphGather layer.  Then, if and only if you specified `error_bars=True` to enable the existing uncertainty estimation, it adds a second dropout layer.

I see two problem with this.  First, we really should have dropout after every convolutional layer.  That's essential for estimating uncertainty.  And without it, the limited amount of dropout it has won't even be very effective for regularization.  Second, the dropout after the GraphGather layer doesn't make sense.  GraphGather defines no variables, so there's nothing for it to regularize.

**Comment by rbharath** [comment]
I'll let @patrickhop add more details, but I believe this was mostly experimental code trying to see what a good uncertainty API might look like for us. I think it would be good to make the modifications you suggest for improving the uncertainty estimate.

**Comment by rbharath** [comment]
@peastman The tutorial is amazing! I actually get the difference between aleatoric and epistemic uncertainty now. This will add a lot of value for our users :+1:

**Comment by patrickhop** [comment]
@peastman That code is from last year and implements (to the best of my knowledge) MC Dropout. If I remember correctly, the formula for computing the error bars here doesn't work well since we don't use weight decay.

My .02 on this work is that we have to think pretty hard about validation. If these uncertainties can withstand a rigorous analysis, this is a huge win for us.

**Comment by peastman** [comment]
Yes, validation is a challenge.  I added a test case that verifies they're at least in the right overall range.  Also see the discussion of it in the tutorial: these uncertainties are really just estimates, and they don't come with any guarantees.

**Comment by patrickhop** [comment]
where can I find the tutorial?

**Comment by peastman** [comment]
Look for "Uncertainty.ipynb" in the list of changed files, then click the "View" button next to it.

**Comment by peastman** [comment]
I didn't calculate the exact coefficient, since they're only expected to be weakly correlated.  The uncertainty just sets an upper limit on the error.  You shouldn't have a small uncertainty and big error, but you'll often have a large uncertainty and small error.

**Comment by peastman** [comment]
Still working on trying to implement this for GraphConvModel.  The next problem I've found is that it overrides both `predict_on_generator()` and `predict_proba_on_generator()`.  Its implementations of them are both buggy and inconsistent with each other.  So as discussed in #1214, I'm trying to get rid of them.  But I don't really understand what they're doing, or why they were overridden in the first place.  Here are the problems and inconsistencies I've found so far.

- `predict_proba_on_generator()` enables dropout during prediction, which will seriously hurt the accuracy of its predictions.  `predict_on_generator()` does not.

- `predict_on_generator()` has some code I don't understand to trim its output.  It first records the number of samples,

```python
n_samples = max(feed_dict[self.membership.out_tensor]) + 1
```

and then trims the output at the end,

```python
for ind, result in enumerate(feed_results):
  # GraphConvModel constantly outputs batch_size number of
  # results, only valid samples should be appended to final results
  results[ind].append(result[:n_samples])
```

But `predict_proba_on_generator()` doesn't do this, so is it actually needed?

- `predict_proba_on_generator()` transposes the output:

```python
if len(result.shape) == 3:
  result = np.transpose(result, axes=[1, 0, 2])
```

but `predict_on_generator()` doesn't do this.  Why do they return different transpositions of the output?

- `predict_proba_on_generator()` ignores its `outputs` arguments and uses `self.outputs` instead.

Can anyone explain what's going on?

**Comment by rbharath** [comment]
I believe that what's going on here is that in order to make graph convs work, TensorFlow requires the input to have a fixed batch size. So, `predict_on_batch` pads its inputs to always be in multiples of batch size. These extra inputs have to be pruned away for the prediction to work correctly. I believe `predict_proba_on_batch` should also be doing this pruning, but isn't doing it for some reason.

**Comment by peastman** [comment]
Do you see any reason we can't make that part of the graph?  It's easy enough to do the same calculation with Tensorflow operators, and then there's no need for special handling in the predict function.

What about the different transpositions?  Any idea what's going on with that?

**Comment by rbharath** [comment]
I think it could probably be made part of the graph. This code has been around for a while, so it can likely be cleaned up. I think we've avoided touching it too much since any structural changes to the graph conv code would require performance testing on MoleculeNet to make sure we didn't accidentally introduce a subtle bug.

I think the transpositions are performed to make the output from the graph conv have the right format.

**Comment by peastman** [comment]
> I think the transpositions are performed to make the output from the graph conv have the right format.

But why are they performed only by `predict_proba_on_generator()`, not `predict_on_generator()`?  This could also easily be made part of the graph, but why does one method transpose and the other not?
